### ISSUE TITLE
Mailgun Template Support

### ISSUE THREAD

**Issue by anstosa** [anymail/django-anymail#155]
The docs claim that
> Mailgun does not offer ESP stored templates, so Anymail’s template_id message attribute is not supported with the Mailgun backend.

but that does not appear to be true: https://documentation.mailgun.com/en/latest/api-templates.html

Am I mistaking either Anymail's or Mailgun's docs?

**Comment by medmunds** [comment]
Hmm... you are not mistaking either docs. 

Mailgun did not support stored templates at the time the Anymail docs were written. This appears to be a new Mailgun feature from… [I guess](https://www.mailgun.com/page/1?s=templates) this [May](https://www.mailgun.com/blog/responsive-email-templates-you-can-get-creative-with)?

Adding stored template support should be relatively simple in Anymail's [Mailgun backend](https://github.com/anymail/django-anymail/blob/v6.1/anymail/backends/mailgun.py#L256). Want to submit a PR?

[Incidentally: I try to follow at least the technical blogs for all supported ESPs, but sometimes miss new feature announcements—particularly when they're disguised as marketing contests.]

**Comment by anstosa** [comment]
Wow that was easy. PR opened: #156

**Comment by medmunds** [comment]
I'm reopening this: I'm looking into a problem with how Mailgun's new templates handle per-recipient merge variables.

It seems like Mailgun's new [stored templates](https://documentation.mailgun.com/en/latest/user_manual.html#templates) don't have access to the `recipient-variables` used with Mailgun's [batch sending](https://documentation.mailgun.com/en/latest/user_manual.html#batch-sending) and on-the-fly templates. (Or at least, you can't get access to those recipient variables using the handlebar syntax you'd expect with the new templates.)
 
Anymail converts its normalized `merge_data` into Mailgun's `recipient-variables`, which is necessary for batch sending. It's not clear to me how you can specify per-recipient variable data for a stored template—all of Mailgun's template examples just use a single recipient. The handlebars variables in the stored templates seem to refer to Mailgun's `v:<variable>` parameters (Anymail's `metadata`), which aren't per-recipient.

I'm going to poke around a bit more, but may have to contact Mailgun to get an answer. (@anstosa if you've already figured this out let me know.)

**Comment by anstosa** [comment]
Oh interesting. We're only doing single recipient sending in our project so I hadn't thought about it

**Comment by medmunds** [comment]
It looks like stored templates *can* still use batch-send `recipient-variables` with the % syntax, like in `Dear %recipient.name%`. But I don't see how to get the `recipient-variables` data from a handlebars template construct (`Dear {{recipient.name}}` doesn't work).

I guess the question is how to expose variables for Mailgun's stored templates in Anymail. Right now, for all other ESPs with stored templates, you use Anymail's [`merge_data`](https://anymail.readthedocs.io/en/stable/sending/templates/#batch-sending-with-merge-data) and `merge_global_data` to supply the template variables. With Mailgun's *stored* templates, you'd need to instead use Anymail's [`metadata`](https://anymail.readthedocs.io/en/stable/sending/anymail_additions/#anymail.message.AnymailMessage.metadata) (Mailgun's `v:<variable>` params), which seems confusing and error-prone.

I'm leaning toward:
1. If a message has a `template_id` and a *single* `to` recipient, pass any `merge_data` for that one recipient (plus any `merge_global_data`) to Mailgun as `X-Mailgun-Variables`. This makes handlebars template variables work as expected for the common single-recipient case: if you're sending only to `to@example.com`, then `merge_data={'to@example.com': {'order_id': 123}}` will be available in the Mailgun template as `{{order_id}}`.
2. If a message has a `template_id` and *multiple* `to` recipients, pass only `merge_global_data` to Mailgun as any `X-Mailgun-Variables`. This allows using stored templates for non-personalized batch sending.
3. If a message has a `template_id` and *multiple* `to` recipients *and* defines per-recipient `merge_data`, raise an UnsupportedFeature error. Although technically supported by Mailgun, it's unlikely this combination of options will behave like the user expects—there's no way to refer to `merge_data` variables in handlebars constructs.
4. In all cases, `merge_data` would *also* continue to be passed as Mailgun `recipient-variables`, so that batch-style `%recipient.var%` references would still work.

(But raising an error in # 3 is the one I'm not so sure about.)

**Comment by anstosa** [comment]
That all seems super reasonable to me, including raising the error in # 3.

**Comment by medmunds** [comment]
OK, so I think I've figured out how to make handlebars template variables work with batch sends, at least for simple string fields. But there's no way to pass arrays or other complex template data with a batch send.

Basically, Anymail should send the per-user merge_data in Mailgun's `recipient-variables` field:

```json
{
  "one@example.com": {"name": "One", "order_id": 123},  
  "two@example.com": {"name": "Two", "order_id": 456}
}
```

And then it needs to pass an `X-Mailgun-Variables` field that defines the handlebars template vars by using`%recipient.<var>%` to pull in the per-recipient value for each variable:

```json
{
   "name": "%recipient.name%",
   "order_id": "%recipient.order_id%"
}
```

As the message is sent to each recipient, Mailgun substitutes in the values for X-Mailgun-Variables, 

The catch is that `X-Mailgun-Variables` has to be valid JSON as sent (*before* the recipient-variables are substituted in), and the only way to do that is with quotes as above—meaning every variable is a string by the time it hits the handlebars template. I can't find any way to end up with an array or object in X-Mailgun-Variables after recipient-variables substitution.

**Comment by mcroni** [comment]
was trying to use the requests library instead of anymail to achieve this and it got this error by doing this as part of the headers
`headers ={
        "X-Mailgun-Variables":{
        "email": "joeydanieldarko@gmail.com",
        "another": "kojo_mcroni@hotmail.com"
    }
`
error
`
requests.exceptions.InvalidHeader: Value for header {X-Mailgun-Variables: {'email': 'joeydanieldarko@gmail.com', 'another': 'kojo_mcroni@hotmail.com'}} must be of type str or bytes, not <class 'dict'>` 
 then i tried using `json.dumps` to get this

`headers ={
        "X-Mailgun-Variables":json.dumps({"email":"love@love.com"})
        }
`
this code doesnt throw any error yet i dont see the variable reflecting in my html template, i use {{ email }} in my template though

**Comment by medmunds** [comment]
@mcroni while waiting for the Anymail update in this issue, you should be able to use Mailgun's handlebars templates with the current version of Anymail. Identify the template in [`esp_extra`](https://anymail.readthedocs.io/en/stable/sending/anymail_additions/#anymail.message.AnymailMessage.esp_extra), and supply the variables in [`metadata`](https://anymail.readthedocs.io/en/stable/sending/anymail_additions/#anymail.message.AnymailMessage.metadata) and/or [`merge_metadata`](https://anymail.readthedocs.io/en/stable/sending/anymail_additions/#anymail.message.AnymailMessage.merge_metadata):

```python
message.esp_extra = {'template': 'my-mailgun-template-name'}
message.metadata = {  # metadata populates Mailgun's handlebars template {{variables}}
    'order_number': '12345',
    'date_shipped': '2019-08-19', 
}
```

Use `metadata` for a single recipient or global data in a batch send, and `merge_metadata` for per-recipient data in a batch send. (This works because Mailgun uses the same underlying mechanism, X-Mailgun-Variables, for both the metadata "user variables" they track with the message *and* for their handlebars "substitution variables".)

[BTW, if you're *not* trying to use Anymail, this probably isn't the best place for your question. Both Mailgun support and Stack Overflow are generally quite helpful with Mailgun technical questions.]

**Comment by medmunds** [comment]
OK, Mailgun support investigated, and confirmed that there's no way to do a *batch send* (per-recipient substitutions) with *complex* handlebars substitution data (arrays and nested objects). But it should work fine for simple string substitutions.

So the new plan for Anymail's Mailgun backend is: 

* If `template_id` is set, then (effectively) combine `merge_data` into `merge_metadata` and `merge_global_data` into `metadata` before sending. This makes Anymail's template merge data available for Mailgun's handlebars variable substitutions in the expected way, including batch sending. (And doesn't break support for Mailgun's other, non-handlebars "template variables".)

* If `template_id` is set, warn about unsupported usage if there are conflicting keys between Anymail's template data (`merge_data`+`merge_global_data`) and metadata (`metadata`+`merge_metadata`), since both have to end up in the same X-Mailgun-Variables.

* Figure out how to document the two different kinds of Mailgun templates and all of Anymail's behavior relating to them. (Including what happens if you try to supply complex data types in `merge_data`/`merge_global_data`, and how to work around that for the single recipient case by manually setting X-Mailgun-Variables.)

**Comment by mcroni** [comment]
I'm actually trying to use anymail with mailgun, will try the provided solution above and give me a feedback soon, thanks for this

**Comment by medmunds** [comment]
Released in Anymail v7.0
